http/agent.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright © The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	defaultPostContentType = "application/octet-stream"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+// Agent is an http agent
+type Agent struct {
+	options *AgentOptions
+	AgentImplementation
+}
+
+// AgentImplementation is the actual implementation of the http calls
+//counterfeiter:generate . AgentImplementation
+type AgentImplementation interface {
+	SendPostRequest(string, []byte, *AgentOptions) (*http.Response, error)
+	SendGetRequest(string, *AgentOptions) (*http.Response, error)
+}
+
+type defaultAgentImplementation struct{}
+
+// AgentOptions has the configurable bits of the agent
+type AgentOptions struct {
+	FailOnHTTPError bool          // Set to true to fail on HTTP Status > 299
+	Timeout         time.Duration // Timeout when fetching URLs
+	Retries         int           // Number of times to retry when errors happen
+	MaxWaitTime     int           // Max waiting time when backing off
+	PostContentType string
+}
+
+// Validate checks the agent options are correct
+func (o *AgentOptions) Validate() error {
+	if o.Retries < 0 {
+		return errors.New("retries has to be equal o greater than 0")
+	}
+	logrus.Infof(
+		"HTTP.Agent options: Timeout: %d - Retries: %d - FailOnHTTPError: %+v",
+		o.Timeout, o.Retries,
+		o.FailOnHTTPError,
+	)
+	return nil
+}
+
+var DefaultAgentOptions = &AgentOptions{
+	FailOnHTTPError: true,
+	Timeout:         0,
+	Retries:         0,
```

Comment:
How about using stronger defaults here? 
```suggestion
	Timeout:         3 * time.Second,
	Retries:         3,
```